### DISCUSSION TITLE
Is Reparenting possible in any way?

### DISCUSSION THREAD

**Discussion by vladnicula** [solidjs/solid#1310] [Q&A]
Hello 👋 I've played around with solid in the past few days and could not find a way to reparent a subtree from one parent to another without rendering the entire subtree. 

I wonder if there is a way to achieve this using solid. 

Imagine having the following hierarchy:
```
<root>
   <nodeA></nodeA>
   <nodeB>
    <Counter />
    <!-- 1000 more Counter component instances -->
     <Counter />
   </nodeB>
</root>
```

And by a user interaction (dragging `nodeB` inside `nodeA`) I want to reparent `nodeB` to be a child of `nodeA` instead of `root`.

If I simply render the content of `nodeB` inside `nodeA` I 
- lose the state of the counters, 
- need to create a lot of html elements that area already created in the page

Ideally, I would like to change the parent of the `nodeB` component.

This does not seem possible.

I tried to cache the results, pasting code below and adding a quick video recording of the behaviour now, but the effect only moves the HTML and loses all interactivity of the components. 

```tsx

export const ComponentInstance = (props: { templateId: string }) => {
    const { templateId } = props
    
    // this will returns the array of children in a signal/store
    // this will not get called after the fist reparenting using the method from 
    // the FOR below 
    const childrenIds = createChildrenIdsSelection(AppState, templateId)
    
    return (
        <div
          style={{
            margin: "16px"
          }}
        >
          <p>Instance for {props.templateId} children: {childrenIds.length}</p>
          <Counter />
          <For each={childrenIds} fallback={null}>
            {(childId) => {
              // return <ComponentInstance templateId={childId} />
              if ( !ComponentInstnceMegaCache.has(childId) ) {
                ComponentInstnceMegaCache.set(childId, <ComponentInstance templateId={childId} />)
              }
              return (
                ComponentInstnceMegaCache.get(childId)
              )
            }}
          </For>
        </div>
      )
  }
  ```
[Video of behaviour using vite ssr svelted starter](https://www.loom.com/share/36fd16f3c36448878f2d68029f8f877e)

This reparenting feature is supported in flutter: https://api.flutter.dev/flutter/widgets/GlobalKey-class.html. It would be really cool if there was a way to have it work in solid as well.

**Comment by atk** [comment]
What about <Portal> and rendering only once the ref for the portal is set using <Show>?

**Reply by vladnicula** [reply]
I don't understand, can you please explain in a more detailed way?

**Reply by atk** [reply]
I just noticed that you want this to happen on drag-and-drop. I would suggest using solid-dnd for that.

**Reply by vladnicula** [reply]
Thanks, but I need to implement this specific thing that uses reparenting. It's not related to drag and drop. It can happen when clicking on an element, or when a server update is pushed. I'm not looking for a black box solution. I am looking specifically for ways to move a DOM element with its entire subhierarchy and internal state from one place of the DOM to some other place.

**Reply by vladnicula** [reply]
I managed to achieve this in React https://3p2xwv.csb.app/

**Reply by atk** [reply]
For react, which re-renders all the time, this requires a completely different approach. Solid usually needs to be manually convinced to render a node somewhere else; usually through a Portal (using a reactive ref signal). You can append the same node to a different parent in the DOM (which removes it from the previous parent), as long as you don't use code that relies on the specific order of elements (for example hydration).

**Comment by ryansolid** [comment]
It's possible. It just can be sort of awkward. The thing is since you can devolve Solid into just reactivity you can pretty much do anything. The problem with the reactive release is because if parent computation re-runs it releases by default, but you can `createRoot` and isolate it. If you step out of declarative zone of like `<For>` components and what not which try to keep track of the shape of things you can can completely do this assuming you remember things like DOM elements can only be attached in a single location. 

Let's consider a thing like Drag and Drop for a moment. The most common pattern for declarative frameworks is to have the drop be captured and then move the element back to its original position, and then tell the framework to move it to the new position so that it's aware of it. If we just left it in the new position the frameworks list indexes would be all off and it'd start moving the wrong elements. Internally the framework doesn't read from the DOM (as that's expensive) and instead keeps its own location information so getting that out of sync is awkward. However if you just didn't use a `<For>` component and decided to take over that section of the DOM manually then no problem, because the framework will never try to move anything and you can do whatever you want.

This doesn't work great with SSR or Hydration per se because the server has no DOM and relies on the shared abstraction, but it is something you could do.

**Reply by vladnicula** [reply]
Thank you for taking the time to answer! 

We are not concerned with SSR, or hydration. I'm actually looking at a way to add a DOM manipulation library in the visual builder of https://play.teleporthq.io/. We used vanillajs in the drag and drop visual builder area and it is getting out of hand. 

Could you point me to anything that would help me understand in more detail your point about the `<For>` and how one might go about handling children separately?  I'm having a hard time wrapping my head around the `createRoot` functionality with my current understanding of solid.

**Reply by vladnicula** [reply]
Oke, after building a signal and effect from scratch from https://dev.to/ryansolid/building-a-reactive-library-from-scratch-1i0p and reading thru the https://indepth.dev/posts/1289/solidjs-reactivity-to-rendering (this one confused me more than it helped 😢 ) I went thru the code of the `createRoot` and `createSignal`. 

I think I need a bit more help understanding how Owner, Listener and Transition work, as from reading the source code it's not clear.

Gonna spend some time reading the unit tests, maybe if figure something out there...
